app/modules/gpio.c

```diff
@@ -16,139 +16,115 @@
 #define HIGH PLATFORM_GPIO_HIGH
 #define LOW PLATFORM_GPIO_LOW
 
-
 #ifdef GPIO_INTERRUPT_ENABLE
 static int gpio_cb_ref[GPIO_PIN_NUM];
-static lua_State* gL = NULL;
-
-void lua_gpio_unref(unsigned pin){
-  if(gpio_cb_ref[pin] != LUA_NOREF){
-    if(gL!=NULL)
-      luaL_unref(gL, LUA_REGISTRYINDEX, gpio_cb_ref[pin]);
-  }
-  gpio_cb_ref[pin] = LUA_NOREF;
-}
 
-void gpio_intr_callback( unsigned pin, unsigned level )
+// This task is scheduled by the ISP if pin_trigger[pin] is true and is used
+// to intitied the Lua-land gpio.trig() callback function
+static void gpio_intr_callback_task (task_param_t param, uint8 priority)
 {
+  unsigned pin = param >> 1;
+  unsigned level = param & 1;
+  UNUSED(priority);
+
   NODE_DBG("pin:%d, level:%d \n", pin, level);
-  if(gpio_cb_ref[pin] == LUA_NOREF)
-    return;
-  if(!gL)
-    return;
-  lua_rawgeti(gL, LUA_REGISTRYINDEX, gpio_cb_ref[pin]);
-  lua_pushinteger(gL, level);
-  lua_call(gL, 1, 0);
+  if(gpio_cb_ref[pin] != LUA_NOREF) {
+    // GPIO callbacks are run in L0 and inlcude the level as a parameter
+    lua_State *L = lua_getstate();
+    NODE_DBG("Calling: %08x\n", gpio_cb_ref[pin]);
+    // The trigger is reset before the callback, as the callback itself might reset the trigger
+    pin_trigger[pin] = true;
+    lua_rawgeti(L, LUA_REGISTRYINDEX, gpio_cb_ref[pin]);
+    lua_pushinteger(L, level);
+    lua_call(L, 1, 0);
+  }
 }
 
 // Lua: trig( pin, type, function )
 static int lgpio_trig( lua_State* L )
 {
-  unsigned type;
-  unsigned pin;
-  size_t sl;
-  
-  pin = luaL_checkinteger( L, 1 );
-  MOD_CHECK_ID( gpio, pin );
-  if(pin==0)
-    return luaL_error( L, "no interrupt for D0" );
-
-  const char *str = luaL_checklstring( L, 2, &sl );
-  if (str == NULL)
-    return luaL_error( L, "wrong arg type" );
-
-  if(sl == 2 && c_strcmp(str, "up") == 0){
-    type = GPIO_PIN_INTR_POSEDGE;
-  }else if(sl == 4 && c_strcmp(str, "down") == 0){
-    type = GPIO_PIN_INTR_NEGEDGE;
-  }else if(sl == 4 && c_strcmp(str, "both") == 0){
-    type = GPIO_PIN_INTR_ANYEDGE;
-  }else if(sl == 3 && c_strcmp(str, "low") == 0){
-    type = GPIO_PIN_INTR_LOLEVEL;
-  }else if(sl == 4 && c_strcmp(str, "high") == 0){
-    type = GPIO_PIN_INTR_HILEVEL;
-  }else{
-    type = GPIO_PIN_INTR_DISABLE;
-  }
-
-  // luaL_checkanyfunction(L, 3);
-  if (lua_type(L, 3) == LUA_TFUNCTION || lua_type(L, 3) == LUA_TLIGHTFUNCTION){
+  unsigned pin = luaL_checkinteger( L, 1 );
+  static const char * const opts[] = {"", "up", "down", "both", "low", "high"};
+  static const int opts_type[] = {
+    GPIO_PIN_INTR_DISABLE, GPIO_PIN_INTR_POSEDGE, GPIO_PIN_INTR_NEGEDGE,
+    GPIO_PIN_INTR_ANYEDGE, GPIO_PIN_INTR_LOLEVEL, GPIO_PIN_INTR_HILEVEL
+    };
+  int type = opts_type[luaL_checkoption(L, 2, "", opts)];
+
+  luaL_argcheck(L, platform_gpio_exists(pin) && pin>0, 1, "Invalid interrupt pin");
```

Comment:
Thanks for the `platform_gpio_exists()`, but why the `pin>0`? If it's due to pin 0 not being interrupt capable, I'd prefer that abstracted out behind a `platform_gpio_intr_capable(pin)` instead.